File: 2352-equal-row-and-column-pairs/solution.cc
struct Trie {
    map<int, Trie*> nxt;
    int cnt; 
};

class Solution {
public:
    int equalPairs(vector<vector<int>>& grid) {
        int ans = 0, n = grid.size();
        auto t = new Trie{};
        for (int i = 0; i < n; ++i) {
            auto curr = t;
            for (int j = 0; j < n; ++j) {
                if (!curr->nxt.count(grid[i][j])) {
                    curr->nxt[grid[i][j]] = new Trie{};
                }
                curr = curr->nxt[grid[i][j]];
            }
            ++curr->cnt;
        }
        for (int j = 0; j < n; ++j) {
            auto curr = t;
            for (int i = 0; i < n; ++i) {
                if (!curr->nxt.count(grid[i][j])) {
                    break;
                }
                curr = curr->nxt[grid[i][j]];
            }
            ans += curr->cnt;
        }

        return ans;
    }
};

File: 2352-equal-row-and-column-pairs/solution.go
type trie struct {
    nxt map[int]*trie
    cnt int
}

func newTrie() *trie {
    return &trie {
        nxt: map[int]*trie{},
        cnt: 0,
    }
}

func equalPairs(grid [][]int) (ans int) {
    t := newTrie()
    for i := range grid {
        curr := t
        for _, v := range grid[i] {
            if _, exists := curr.nxt[v]; !exists {
                curr.nxt[v] = newTrie()
            }
            curr = curr.nxt[v]
        }
        curr.cnt++
    }

    n := len(grid)
    for j := 0; j < n; j++ {
        curr := t
        for i := 0; i < n; i++ {
            if _, exists := curr.nxt[grid[i][j]]; !exists {
                break
            }
            curr = curr.nxt[grid[i][j]]
        }
        ans += curr.cnt
    }

    return ans
}

